dashboard/lib/policies/script_activity.rb

```diff
@@ -0,0 +1,25 @@
+class Policies::ScriptActivity
+  def self.completed?(user, script)
+    return false unless user
+    return false unless script
+    user_script = user.user_scripts.find_by(script: script)
+    return false unless user_script
+    !!user_script.completed_at || user.completed_progression_levels?(script)
+  end
+
+  def self.passed_any_level?(user, script)
+    user_levels = UserLevels.
+      where(user: user, script: script).
+      index_by(&:level_id)
+
+    script.script_levels.detect do |script_level|
+      user_level = user_levels[script_level.level_id]
+      is_passed = (user_level && user_level.passing?)
+      script_level.valid_progression_level? && is_passed
+    end
+  end
+
+  def self.not_started?(user, script)
+    !completed?(user, script) && !passed_any_level?(user, script)
```

Comment:
this logic would become slightly cleaner if we instantiated this object as discussed earlier

Reply:
The problem I've run into before is that the internal logic of the Policy Object becomes cleaner, but calling it turns from 

```ruby
Policies::ScriptActivity.not_started?(user, script)
```

into

```ruby
Policies::ScriptActivity.new(user, script).not_started?
```

Which is only slightly less clean for this case, but is a pattern that can become unwieldy in other cases. See https://github.com/code-dot-org/code-dot-org/pull/30752 for more discussion.

Reply:
Good point -- it's a toss up which is less syntax, so I'm fine with leaving these as class methods.